File: src/lib/occupiedTerritories.ts
/** Occupied territory guardrails — mirrors bot/occupied_territories.py bbox checks. */

export type OccupiedZoneId =
  | "crimea"
  | "donetsk-occupied"
  | "luhansk-occupied"
  | "zaporizhzhia-occupied-south"
  | "kherson-occupied-east-bank"

/** (south, west, north, east) WGS84 — simplified product guardrails. */
export const OCCUPIED_BBOXES: ReadonlyArray<{ id: OccupiedZoneId; bbox: readonly [number, number, number, number] }> = [
  { id: "crimea", bbox: [44.3, 32.5, 46.2, 36.8] },
  { id: "donetsk-occupied", bbox: [47.0, 37.5, 49.8, 40.2] },
  { id: "luhansk-occupied", bbox: [48.0, 38.8, 50.1, 40.5] },
  { id: "zaporizhzhia-occupied-south", bbox: [46.0, 34.8, 47.4, 36.8] },
  { id: "kherson-occupied-east-bank", bbox: [46.0, 32.5, 47.15, 35.0] },
]

function inBbox(lat: number, lng: number, bbox: readonly [number, number, number, number]): boolean {
  const [south, west, north, east] = bbox
  return lat >= south && lat <= north && lng >= west && lng <= east
}

export function isOccupiedCoordinates(lat: number | null | undefined, lng: number | null | undefined): boolean {
  if (lat == null || lng == null || !Number.isFinite(lat) || !Number.isFinite(lng)) return false
  return OCCUPIED_BBOXES.some(({ bbox }) => inBbox(lat, lng, bbox))
}

export function occupiedZoneName(lat: number | null | undefined, lng: number | null | undefined): OccupiedZoneId | null {
  if (lat == null || lng == null || !Number.isFinite(lat) || !Number.isFinite(lng)) return null
  for (const zone of OCCUPIED_BBOXES) {
    if (inBbox(lat, lng, zone.bbox)) return zone.id
  }
  return null
}

export const OCCUPIED_PICK_MESSAGE = "Ця територія тимчасово окупована. Оберіть точку на підконтрольній Україні."

File: bot/occupied_territories.py
"""Occupied territory checks per de-facto UA government control lines (2022+).

Used to block orders/geocoding and hide directory POIs in temporarily occupied areas.
See docs/OCCUPIED_TERRITORIES.md for the zone list and rationale.

Zones are simplified axis-aligned bounding boxes — conservative enough for product
guardrails; refine polygons when higher precision is needed.
"""

from __future__ import annotations

from typing import Any

# Each zone: (south, west, north, east) in WGS84 degrees.
OCCUPIED_BBOXES: tuple[tuple[str, tuple[float, float, float, float]], ...] = (
    ("crimea", (44.30, 32.50, 46.20, 36.80)),
    ("donetsk-occupied", (47.00, 37.50, 49.80, 40.20)),
    ("luhansk-occupied", (48.00, 38.80, 50.10, 40.50)),
    ("zaporizhzhia-occupied-south", (46.00, 34.80, 47.40, 36.80)),
    ("kherson-occupied-east-bank", (46.00, 32.50, 47.15, 35.00)),
)


def _in_bbox(lat: float, lng: float, bbox: tuple[float, float, float, float]) -> bool:
    south, west, north, east = bbox
    return south <= lat <= north and west <= lng <= east


def is_occupied_coordinates(lat: float | None, lng: float | None) -> bool:
    """True when point lies inside any occupied zone."""
    if lat is None or lng is None:
        return False
    try:
        lat_f = float(lat)
        lng_f = float(lng)
    except (TypeError, ValueError):
        return False
    if not (-90 <= lat_f <= 90 and -180 <= lng_f <= 180):
        return False
    return any(_in_bbox(lat_f, lng_f, bbox) for _, bbox in OCCUPIED_BBOXES)


def occupied_zone_name(lat: float | None, lng: float | None) -> str | None:
    if lat is None or lng is None:
        return None
    try:
        lat_f = float(lat)
        lng_f = float(lng)
    except (TypeError, ValueError):
        return None
    for name, bbox in OCCUPIED_BBOXES:
        if _in_bbox(lat_f, lng_f, bbox):
            return name
    return None


def filter_non_occupied_providers(providers: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Drop directory/dispatch providers whose map pin is in occupied territory."""
    kept: list[dict[str, Any]] = []
    for provider in providers:
        location = provider.get("location") if isinstance(provider.get("location"), dict) else {}
        lat = location.get("lat")
        lng = location.get("lng")
        if is_occupied_coordinates(lat, lng):
            continue
        kept.append(provider)
    return kept
